R-package/R/utils.R

```diff
@@ -250,16 +250,38 @@ lgb.check.obj <- function(params, obj) {
 
 }
 
+# [description] Take any character values from eval and store them
+#               in params$metric. This has to account for the fact that
+#               `eval` could be a character vector, a function, a list of functions,
+#               or a list with a mix of strings and functions
 lgb.check.eval <- function(params, eval) {
 
   # Check if metric is null, if yes put a list instead
   if (is.null(params$metric)) {
     params$metric <- list()
   }
 
-  # If 'eval' is a list or character vector, store it in 'metric'
-  if (is.character(eval) || identical(class(eval), "list")) {
-    params$metric <- append(params$metric, eval)
+  # if 'eval' is a character vector or list, find the character
+  # elements and add them to 'metric'
+  if (!is.function(eval)) {
+    for (i in seq_along(eval)) {
+      element <- eval[[i]]
+      if (is.character(element)) {
+        print(paste0("Adding '", element, "' to list of metrics"))
+        params$metric <- append(params$metric, element)
+      }
+    }
+  }
+
+  # If more than one character metric was given, then "None" should
+  # not be included
+  if (length(params$metric) > 1L) {
+    params$metric <- Filter(
+        f = function(metric) {
+          metric != "None"
+        }
+        , x = params$metric
+    )
   }
```

Comment:
It might be better to check for the following elements: `c("None" "na", "null", "custom")` according to https://github.com/microsoft/LightGBM/blob/master/docs/Parameters.rst#metric

Perhaps also any `anyNA`, but it depends on doc interpretation.

But if we do this change, it would also require the change there: 

* https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/engine.py#L76
* https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/engine.py#L444